internal/controller/bmcsettings_controller.go

```diff
@@ -0,0 +1,861 @@
+	ctx context.Context,
+	log logr.Logger,
+	bmcSetting *metalv1alpha1.BMCSettings,
+	BMC *metalv1alpha1.BMC,
+) {
+	log.V(1).Info("Handle failed setting update with no maintenance reference")
+	// todo: revisit this logic to either create maintenance if not present, put server in Error state on failed bmc settings maintenance
+	log.V(1).Info("Failed to update BMC setting", "ctx", ctx, "bmcSetting", bmcSetting, "BMC", BMC)
+}
+
+func (r *BMCSettingsReconciler) getBMCVersionAndSettingsDifference(
+	ctx context.Context,
+	log logr.Logger,
+	bmcSetting *metalv1alpha1.BMCSettings,
+	BMC *metalv1alpha1.BMC,
+	bmcClient bmc.BMC,
+) (currentBMCVersion string, diff redfish.SettingsAttributes, err error) {
+	keys := slices.Collect(maps.Keys(bmcSetting.Spec.SettingsMap))
+
+	currentSettings, err := bmcClient.GetBMCAttributeValues(ctx, BMC.Spec.BMCUUID, keys)
+	if err != nil {
+		log.V(1).Info("Failed to get with BMC setting", "error", err)
+		return currentBMCVersion, diff, fmt.Errorf("failed to get BMC settings: %w", err)
+	}
+
+	diff = redfish.SettingsAttributes{}
+	var errs []error
+	for key, value := range bmcSetting.Spec.SettingsMap {
+		res, ok := currentSettings[key]
+		if ok {
+			switch data := res.(type) {
+			case int:
+				intvalue, err := strconv.Atoi(value)
+				if err != nil {
+					log.V(1).Info("Failed to check type for", "Setting name", key, "setting value", value, "error", err)
+					errs = append(errs, fmt.Errorf("failed to check type for name %v; value %v; error: %v", key, value, err))
+					continue
+				}
+				if data != intvalue {
+					diff[key] = intvalue
+				}
+			case string:
+				if data != value {
+					diff[key] = value
+				}
+			case float64:
+				floatvalue, err := strconv.ParseFloat(value, 64)
+				if err != nil {
+					log.V(1).Info("Failed to check type for", "Setting name", key, "setting value", value, "error", err)
+					errs = append(errs, fmt.Errorf("failed to check type for name %v; value %v; error: %v", key, value, err))
+				}
+				if data != floatvalue {
+					diff[key] = floatvalue
+				}
+			}
+		} else {
+			diff[key] = value
+		}
+	}
+
+	if len(errs) > 0 {
+		return currentBMCVersion, diff, fmt.Errorf("failed to find diff for some BMC settings: %v", errs)
+	}
+
+	// fetch the current BMC version from the server bmc
+	currentBMCVersion, err = bmcClient.GetBMCVersion(ctx, BMC.Spec.BMCUUID)
+	if err != nil {
+		return currentBMCVersion, diff, fmt.Errorf("failed to load BMC version: %w for BMC %v", err, BMC.Name)
+	}
+
+	return currentBMCVersion, diff, nil
+}
+
+func (r *BMCSettingsReconciler) checkForRequiredPowerStatus(
+	BMC *metalv1alpha1.BMC,
+	powerState metalv1alpha1.BMCPowerState,
+) bool {
+	return BMC.Status.PowerState == powerState
+}
+
+func (r *BMCSettingsReconciler) checkIfMaintenanceGranted(
+	ctx context.Context,
+	log logr.Logger,
+	bmcSetting *metalv1alpha1.BMCSettings,
+	bmcClient bmc.BMC,
+) bool {
+	if bmcSetting.Spec.ServerMaintenanceRefList == nil {
+		return true
+	}
```

Comment:
Returning `true` on an empty list, could be an easy source for bugs, because it requires that `requestMaintenances()` is called on all code paths beforehand. Could `false` be returned instead? (A BMC without a server, seems somewhat unrealistic to me.)